File: migen/pwm_only_v2.py
from migen import *
from migen.fhdl import verilog

class pwm_only_v2(Module) :
    def __init__(self):

        self.wr_data = Signal(32)
        self.cs = Signal()
        self.wr_n = Signal()
        self.addr = Signal(2)
        self.rd_data = Signal(32)
        self.pwm_out = Signal(8)

        #registers
        self.div = Signal(32)
        self.duty_a = Signal(32)
        self.duty_b = Signal(32)
        self.counter = Signal(32)
        self.off_a = Signal()
        self.off_b = Signal()

        #chip select and address decoder
        self.div_en = Signal()
        self.duty_en = Signal()
        self.duty_a_en = Signal()
        self.duty_b_en = Signal()

        self.comb += self.div_en.eq(self.cs & ~self.wr_n & ~self.addr[1])
        self.comb += self.duty_en.eq(self.cs & ~self.wr_n & self.addr[1])
        self.comb += self.duty_a_en.eq(self.duty_en & ~self.addr[0])
        self.comb += self.duty_b_en.eq(self.duty_en & self.addr[0])

        #register write
        self.sync += If(self.div_en, self.div.eq(self.wr_data))
        self.sync += If(self.duty_a_en, self.duty_a.eq(self.wr_data))
        self.sync += If(self.duty_b_en, self.duty_b.eq(self.wr_data))

        #register read
        self.comb += If(self.addr[1] == 0, 
                        self.rd_data.eq(self.div)
                        ).Else(
                            If(self.addr[0] == 0,
                               self.rd_data.eq(self.duty_a)
                               ).Else(
                                   self.rd_data.eq(self.duty_b)
                               )
                        )
        
        #//////PWM logic//////
        
        #PWM counter
        self.sync += If(self.counter >= self.div, 
                        self.counter.eq(0)
                        ).Else(
                            self.counter.eq(self.counter + 1)
                        )

        #PWM compare
        self.sync += If(self.counter >= self.duty_a,
                        self.off_a.eq(1)
                        ).Elif(self.counter == 0, 
                               self.off_a.eq(0)
                               ).Else(self.off_a.eq(self.off_a))
        
        self.sync += If(self.counter >= self.duty_b,
                        self.off_b.eq(1)
                        ).Elif(self.counter == 0, 
                               self.off_b.eq(0)
                               ).Else(self.off_b.eq(self.off_b))
        
        self.comb += self.pwm_out.eq((Replicate(~self.off_a, 4) << 4) | Replicate(~self.off_b, 4))

if (__name__ == "__main__"):
    m = pwm_only_v2()
    print(verilog.convert(m, {m.wr_data, m.cs, m.wr_n, m.addr, m.rd_data, m.pwm_out}))

File: chisel/pwm_only_v2.scala
//> using scala "2.13.12"
//> using dep "org.chipsalliance::chisel:6.6.0"
//> using plugin "org.chipsalliance:::chisel-plugin:6.6.0"
//> using options "-unchecked", "-deprecation", "-language:reflectiveCalls", "-feature", "-Xcheckinit", "-Xfatal-warnings", "-Ywarn-dead-code", "-Ywarn-unused", "-Ymacro-annotations"

import chisel3._
import chisel3.util._
// _root_ disambiguates from package chisel3.util.circt if user imports chisel3.util._
import _root_.circt.stage.ChiselStage

class new_pwm_v2 extends Module {

  val wr_data = IO(Input(UInt(32.W)))
  val cs      = IO(Input(Bool()))
  val wr_n    = IO(Input(Bool()))
  val addr    = IO(Input(UInt(2.W)))
  val rd_data = IO(Output(UInt(32.W)))
  val pwm_out  = IO(Output(UInt(8.W)))

  //Registers and wires
  val div = RegInit(0.U(32.W))
  val duty_a = RegInit(0.U(32.W))
  val duty_b = RegInit(0.U(32.W))
  val counter = RegInit(0.U(32.W))
  val off_a = RegInit(0.U(1.W))
  val off_b = RegInit(0.U(1.W))
  val div_en = Wire(Bool())
  val duty_en = Wire(Bool())
  val duty_a_en = Wire(Bool())
  val duty_b_en = Wire(Bool())
  
  // Avalon slave interface
  div_en := cs & ~wr_n & ~addr(1)
  duty_en := cs & ~wr_n & addr(1)
  duty_a_en := duty_en & ~addr(0)
  duty_b_en := duty_en &  addr(0)

  //register write
  when (div_en) {
    div := wr_data
  }
  when (duty_a_en) {
    duty_a := wr_data
  }
  when (duty_b_en) {
    duty_b := wr_data
  }

  // register read
  rd_data := Mux(~addr(1), 
                  div, Mux(~addr(0), 
                            duty_a, duty_b))
  
  //PWM Counter
  counter := Mux(counter >= div, 0.U, counter + 1.U)
  
  //PWM Compare
  off_a := Mux(counter >= duty_a, 
                1.U, Mux(counter === 0.U,
                          0.U, off_a))
  
  off_b := Mux(counter >= duty_b, 
                1.U, Mux(counter === 0.U,
                          0.U, off_b))

  pwm_out := Cat(Fill(4, ~off_a), Fill(4, ~off_b))
}



object Main extends App {
  println(
    ChiselStage.emitSystemVerilog(
      gen = new new_pwm_v2,
      firtoolOpts = Array("-disable-all-randomization", "-strip-debug-info")
    )
  )
}
